### PR TITLE
Add documentation and tests for builtins.go

### PR DESCRIPTION
**Pull request by arsham**
It changes `TypeCmd` and `EnvCmd` to functions as they are not changing the state of `BuiltinCmds` values. `ExecCmd` probably should be a function too but it was used [here](https://github.com/KurokuLabs/margo/blob/master/mg/cmd.go#L125) so I didn't touch it.

### CODE REVIEW

File: mg/builtins.go [hunk 1]
```diff
@@ -3,9 +3,10 @@ package mg
 import (
 	"bytes"
 	"fmt"
-	"margo.sh/mgutil"
 	"os"
 	"sort"
+
```
Comment (DisposaBoy) [inline mg/builtins.go]: see the `gofmt` comment on the other PR

File: mg/builtins.go [hunk 2]
```diff
@@ -31,14 +32,19 @@ func (bcl BultinCmdList) Lookup(name string) (cmd BultinCmd, found bool) {
 	panic("internal error: the `.exec` BuiltinCmd is not defined")
 }
 
+// BuiltinCmds is used to add a list of predefined commands to the state. It
```
Comment (DisposaBoy) [inline mg/builtins.go]: I'd say something like: `BuiltinCmds implements various builtin commands`. Once you understand how reducers and the state works, it should all be clear how this works so no need to mention it.

Although saying that, it's probably better to just not export this type. `Builtins` is meant to be used like a namespace (so TypeCmd, etc. should be methods instead of global functins) and the type isn't/shouldn't be useful to users outside the package.

File: mg/builtins.go [hunk 3]
```diff
@@ -31,14 +32,19 @@ func (bcl BultinCmdList) Lookup(name string) (cmd BultinCmd, found bool) {
...
+// implements Reducer interface.
```
Comment (DisposaBoy) [inline mg/builtins.go]: I'd also not bother mentioning which types implement which interfaces, we should just improve the tools instead.

File: mg/builtins.go [hunk 4]
```diff
@@ -31,14 +32,19 @@ func (bcl BultinCmdList) Lookup(name string) (cmd BultinCmd, found bool) {
...
 type BuiltinCmds struct{}
 
+// ExecCmd spawns the given command in a goroutine and returns its State right
```
Comment (DisposaBoy) [inline mg/builtins.go]: I'd just say something like `ExecCmd implements the \`.exec\` builtin`. It's ran in a reducer and the command can take a while to run so it's already implied that it's non-blocking unless documented otherwise.

File: mg/builtins.go [hunk 5]
```diff
@@ -31,14 +32,19 @@ func (bcl BultinCmdList) Lookup(name string) (cmd BultinCmd, found bool) {
...
+// away. If the name is `.exec`, it will use its first argument as the `Name`
```
Comment (DisposaBoy) [inline mg/builtins.go]: No need to add comments like these, they're just stating what the implementation does and you'd then need to change the docs every time the implementation changes.

File: mg/builtins.go [hunk 6]
```diff
@@ -74,12 +83,14 @@ func (bc BuiltinCmds) TypeCmd(bx *BultinCmdCtx) *State {
...
 
-func (bc BuiltinCmds) EnvCmd(bx *BultinCmdCtx) *State {
+// EnvCmd finds all environment variables corresponding to bx.Args into the
+// bx.Output buffer.
+func EnvCmd(bx *BultinCmdCtx) *State {
 	buf := &bytes.Buffer{}
 	names := bx.Args
 	if len(names) == 0 {
 		names = make([]string, 0, len(bx.Env))
-		for k, _ := range bx.Env {
```
Comment (DisposaBoy) [inline mg/builtins.go]: Omitting `_` in func args and receivers is fine (I prefer the consistency, but have no strong opionion on it), but please keep them for ranges. It removes any confusion about what the variable holds and avoids bugs esp. considering channels returns the value by default.

File: mg/builtins.go [hunk 7]
```diff
@@ -92,14 +103,17 @@ func (bc BuiltinCmds) EnvCmd(bx *BultinCmdCtx) *State {
...
+		BultinCmd{Name: ".env", Desc: "List env vars", Run: EnvCmd},
 		BultinCmd{Name: ".exec", Desc: "Run a command through os/exec", Run: bc.ExecCmd},
-		BultinCmd{Name: ".type", Desc: "Lists all builtins or which builtin handles a command", Run: bc.TypeCmd},
+		BultinCmd{Name: ".type", Desc: "Lists all builtins or which builtin handles a command", Run: TypeCmd},
 	}
 }
 
+// Reduce will return a copy of the mx.State while adding the predefined
+// commands to its builtins, if the mx.Action is a RunCmd.
 func (bc BuiltinCmds) Reduce(mx *Ctx) *State {
```
Comment (DisposaBoy) [inline mg/builtins.go]: I'd wprobably ord this like `Reduce adds the list of predefined builtins for the RunCmd action`. It's already implied that if the state changes, the reducer returns a new state object.

File: mg/builtins_test.go [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+			t.Errorf("buf.String() = (%s); want (%s) in it", out, item.Name)
+		}
+		if !strings.Contains(out, item.Desc) {
+			t.Errorf("buf.String() = (%s); want (%s) in it", out, item.Desc)
+		}
+	}
+}
+
+func buildBultinCmdCtx(cmds mg.BultinCmdList, args []string, envMap mg.EnvMap, buf io.Writer) *mg.BultinCmdCtx {
+	ctx := &mg.Ctx{
```
Comment (DisposaBoy) [inline mg/builtins_test.go]: It's safest to use the `newXXX()` functions when available. They often make sure things are initialised correctly.
e.g. `newCtx` initialises things like `Store` because all the code assumes that if you give it a `*Ctx` then it can access all the fields.

File: mg/builtins_test.go [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+		State: &mg.State{
+			BuiltinCmds: cmds,
+		},
+	}
+	ctx.Env = envMap
+	return &mg.BultinCmdCtx{
```
Comment (DisposaBoy) [inline mg/builtins_test.go]: Same as the comment about `newCtx`, it's better to use `NewBultinCmdCtx()` here.

File: mg/builtins.go (hunk 1 above)
Comment (arsham) [reply mg/builtins.go]: :+1:

File: mg/builtins_test.go (hunk 8 above)
Comment (DisposaBoy) [reply mg/builtins_test.go]: Good point. I will export a function to create a new ctx. I think you're already out of sync with the repo, but it shouldn't cause any merge conflicts

Comment (DisposaBoy) [reply mg/builtins_test.go]: I just pushed it. Use it like

	ag, _ := mg.NewAgent(mg.AgentConfig{})
	ctx, done := ag.NewCtx(nil)
	defer close(done)

Comment (DisposaBoy) [reply mg/builtins_test.go]: BTW, if I don't repond the some comments, ping me. This github ui is hard to follow and I think I might be missing some of your comments.

Comment (arsham) [reply mg/builtins_test.go]: :+1:

**Comment by arsham** [comment]
I prefer squashing these commits to keep the merge history clean.

File: mg/builtins.go [hunk 10]
```diff
@@ -31,14 +31,16 @@ func (bcl BultinCmdList) Lookup(name string) (cmd BultinCmd, found bool) {
...
 type BuiltinCmds struct{}
 
+// ExecCmd implements the `.exec` builtin.
 func (bc BuiltinCmds) ExecCmd(bx *BultinCmdCtx) *State {
-	go bc.execCmd(bx)
+	go execCmd(bx)
 	return bx.State
 }
 
-func (bc BuiltinCmds) execCmd(bx *BultinCmdCtx) {
```
Comment (DisposaBoy) [inline mg/builtins.go]: please make these functions methods again

Comment (arsham) [reply mg/builtins.go]: Will do.
